### ISSUE TITLE
Feature Request: [Keras] Add _keras_history to tensor in Eager Mode

### ISSUE THREAD

**Issue by howl-anderson** [tensorflow/tensorflow#37119]
For some losses such as CRF loss, it's computation does not only rely on `y_true` and `y_pred`, it must access the layer's internal variable.

 In such a situation, `_keras_history` will be greate helpful.  The loss function can use `_keras_history` of `y_pred` to get the layer instance then get the variables. 

But in eager mode, the output tensor does not have `_keras_history`, which makes such losses not easy to implement in Keras without a custom patch to layers or custom models.

related to tensorflow/addons#377

**Comment by omalleyt12** [comment]
@howl-anderson Thanks for the issue!

This isn't something we can support in eager mode. KerasHistory is only set on a Tensor during the Functional API construction phase.

For losses that require variables, you can use `add_loss` inside a custom `Layer`:

```python
class LossLayer(tf.keras.layers.Layer):
  def call(self, x, y):
    ...
    self.add_loss(my_loss(x, y, self.variable))
   return x

x_inputs = tf.keras.Input(...)
y_inputs = tf.keras.Input(...)
outputs = LossLayer()(x_inputs, y_inputs)
model = tf.keras.Model([x_inputs, y_inputs], outputs)
```
